### PR TITLE
E2E coverage on Snap Accounts and Snap Simple Keyring

### PR DESCRIPTION
**Pull request by HowardBraham**
### Covers the following tests from the test plan
- Installation of snap and creation of account ✅
- Sign and broadcast a p2p send transaction to testnet using a snap account ✅
    - Reject p2p send transaction to testnet using snap account ✅
- Overall signing using a snap account from the Test Dapp (use all signing methods Excluding ETH SIGN- will be added later) ✅
    - Reject signature using a snap account from the Test Dapp (use ~~all~~ most signing methods, but the preference is Sign Typed Data V3 and V4) ✅
    - Disconnect, connect a snap account from the Test Dapp to execute signing methods ✅

The suite at `test/e2e/accounts/test-snap-accounts.spec.js` runs 20 different E2E tests, some of them very similar to each other.  I used a `foreach` technique to loop through `it` tests, described here: https://muratcorlu.com/mocha-js-equivalent-of-it-each/

## **Manual testing steps**

To run the E2E tests locally, use
`yarn test:e2e:single test/e2e/accounts/test-snap-accounts.spec.js --browser=firefox --debug`

## **Related issues**

Part of MetaMask/accounts-planning#44

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained:
  - [x] What problem this PR is solving.
  - [x] How this problem was solved.
  - [x] How reviewers can test my changes.
- [x] I’ve indicated what issue this PR is linked to: Fixes #???
- [x] I’ve included tests if applicable.
- [x] I’ve documented any added code.
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
- [x] I’ve properly set the pull request status:
  - [ ] In case it's not yet "ready for review", I've set it to "draft".
  - [x] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [X] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [X] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: test/e2e/snaps/test-snap-manageAccount.spec.js [hunk 1]
```diff
@@ -1,106 +1,452 @@
...
+
+        await installSnapSimpleKeyring(driver, isAsyncFlow);
+
+        const newPublicKey = await makeNewAccountAndSwitch(driver);
+
+        // open the Test Dapp and connect Account 2 to it
+        await connectAccountToTestDapp(driver);
+
+        // do #signTypedDataV3
+        await signData(driver, '#signTypedDataV3', newPublicKey, flowType);
```
Comment (plasmacorral) [inline test/e2e/snaps/test-snap-manageAccount.spec.js:154]: Do we need an assertion here to verify that this worked and to confirm that the signing address is verified against the public address used?

File: test/e2e/snaps/test-snap-manageAccount.spec.js [hunk 2]
```diff
@@ -1,106 +1,452 @@
...
+
+        // disconnect from the Test Dapp
+        await disconnectFromTestDapp(driver);
+
+        // reconnect Account 2 to the Test Dapp
+        await connectAccountToTestDapp(driver);
+
+        // do #signTypedDataV4
+        await signData(driver, '#signTypedDataV4', newPublicKey, flowType);
```
Comment (plasmacorral) [inline test/e2e/snaps/test-snap-manageAccount.spec.js:163]: Do we need an assertion here to verify that this worked and to confirm that the signing address is verified against the public address used?

File: test/e2e/accounts/test-snap-accounts.spec.js [hunk 3]
```diff
@@ -0,0 +1,452 @@
...
+    // click the parent of #use-sync-flow-toggle (trying to click the element itself gives "ElementNotInteractableError: could not be scrolled into view")
+    await driver.clickElement({
+      xpath: '//input[@id="use-sync-flow-toggle"]/..',
+    });
+  }
+
+  /**
+   * @param {Driver} driver
+   */
+  async function importKeyAndSwitch(driver) {
```
Comment (plasmacorral) [inline test/e2e/accounts/test-snap-accounts.spec.js:311]: NIT: This should probably be done with a fixture going forward.

File: test/e2e/accounts/test-snap-accounts.spec.js [hunk 4]
```diff
@@ -0,0 +1,452 @@
...
+  async function importPrivateKeyAndTransfer1ETH(driver, flowType) {
+    const isAsyncFlow = flowType !== 'sync';
+
+    await installSnapSimpleKeyring(driver, isAsyncFlow);
+    await importKeyAndSwitch(driver);
+
+    // send 1 ETH from Account 2 to Account 1
+    await sendTransaction(driver, PUBLIC_KEY, 1, isAsyncFlow);
+
+    if (isAsyncFlow) {
```
Comment (plasmacorral) [inline test/e2e/accounts/test-snap-accounts.spec.js:181]: NIT: If statements are generally discouraged in e2e due to adding complexity and variability: https://docs.google.com/document/d/1eQyeScSZQlzGLapef_8PANkU_ObNPJcuMCP040T2gfM/edit#heading=h.njqh4zz4ol4w

**Review by plasmacorral — APPROVED** [review]
Still missing Eth Sign, but have other signing methods otherwise covered here.  Can add that in a future change along with other account/snap removal tests.

File: package.json [hunk 5]
```diff
@@ -16,6 +16,7 @@
     "build": "yarn lavamoat:build",
     "build:dev": "node development/build/index.js",
     "start:test": "SEGMENT_HOST='https://api.segment.io' SEGMENT_WRITE_KEY='FAKE' SENTRY_DSN_DEV=https://fake@sentry.io/0000000 PORTFOLIO_URL=http://127.0.0.1:8080 yarn build:dev testDev",
+    "start:test:flask": "SEGMENT_HOST='https://api.segment.io' SEGMENT_WRITE_KEY='FAKE' SENTRY_DSN_DEV=https://fake@sentry.io/0000000 PORTFOLIO_URL=http://127.0.0.1:8080 BLOCKAID_FILE_CDN=storage.googleapis.com/ppom-mock-cdn yarn build:dev testDev --build-type flask --apply-lavamoat=false --snow=false",
```
Comment (gantunesr) [inline package.json:19]: I thought this was added in another PR

**Review by gantunesr — APPROVED** [review]
LGTM. Just minor questions

File: test/e2e/fixture-builder.js [hunk 6]
```diff
@@ -580,7 +580,7 @@ class FixtureBuilder {
     return this;
   }
 
-  withPermissionControllerConnectedToTestDapp() {
+  withPermissionControllerConnectedToTestDapp(restrictReturnedAccounts = true) {
```
[the remark this answers is not in the document]
Comment (gantunesr) [reply test/e2e/fixture-builder.js:583]: nvm

File: package.json (hunk 5 above)
Comment (HowardBraham) [reply package.json:19]: I think I may have stuck it in the draft of another PR, but it never got merged.

**Comment by HowardBraham** [comment]
> Still missing Eth Sign

Oh whoops, I got stuck here.  I think a previous version of the Snap didn't allow Eth Sign at all, but now you have to enable in Settings.
image

**Pull request merged**